docs/RFCS/stateless_replica_relocation.md

```diff
@@ -0,0 +1,232 @@
+- Feature Name: Stateless Replica Relocation
+- Status: draft
+- Start Date: 2015-08-19
+- RFC PR:
+- Cockroach Issue: #620
+
+# Summary
+A relocation is, conceptually, the transfer of a single replica from one store to
+another. However, the implementation is necessarily the combination of two operations:
+
+1. Creating a new replica for a range on a new store.
+2. Removing a replica of the same range from a different store.
+
+For example, by creating a replica on store Y and then removing a replica from
+store X, you have in effect moved the replica from X to Y.
+
+This PR is suggesting an overall architectural design goal: that the decision to
+make any individual operation (either a create or a remove) should be **stateless**.
+In other words, the second operation in a replica relocation should not depend on a
+specific invocation of the first operation.
+
+# Motivation
+For an assortment of reasons, Cockroach will often need to relocate the replicas
+of a range. Most immediately, this is needed for repair (when a store dies and
+its replicas are no longer usable) and rebalance (relocating replicas on
+overloaded stores to stores with excess capacity).
+
+In order to retain the integrity of a range's raft consensus group, a relocation
+must be expressed as a combination of two operations:
+
+1. Creating a new replica for a range on a new store.
+2. Removing a replica of the same range from a different store.
+
+These operations can happen in either order as long as quorum is maintained in
+the range's raft group after each individual operation, although one ordering may be
+preferred over another.
+
+Expressing a specific relocation (i.e. "move replica from store X to store Y")
+would require maintaining some persistent state to link the two operations
+involved.  Storing that state presents a number of issues: where is it stored,
+in memory or on disk? If it's in memory, does it have to be replicated through
+raft or is it local to one node?  If on disk, can the persistent state become
+stale? How do you detect conflicts between two relocation operations initiated
+from different places?
+
+This RFC suggests that no such relocation state should be persisted. Instead, a
+system that wants to initiate a relocation will perform only the first
+operation, and a different system will detect the need for a complementary
+operation and perform it. A relocation is thus completed without any state being
+exchanged between those two systems.
+
+By eliminating the need to persist any data about in-progress relocation
+operations, the overall system is dramatically simplified.
+
+# Detailed design
+The implementation involves a few pieces:
+
+1. Each range must have a persisted target replication state. This does not
+   prescribe specific replica locations; it suggests only a desired number of
```

Comment:
I don't know if it's law, the cluster works with 1 node even though it desires 3 replicas for everything.